### ISSUE TITLE
NodeJS (multer upload)

### ISSUE THREAD

**Issue by loicmiramon** [LinusU/fs-temp#11]
Hello,

I am learning nodeJS and I am having a problem on multer here is the error, if you have a solution I am a taker ! thank's

The "path" argument must be of type string or an instance of Buffer or URL. Received nul

<img width="719" alt="Capture d’écran 2022-08-04 à 17 55 20" src="https://user-images.githubusercontent.com/45937748/182892078-8d76f737-63cd-4a66-a2ee-8dbe6a6b051a.png">
<img width="796" alt="Capture d’écran 2022-08-04 à 17 55 47" src="https://user-images.githubusercontent.com/45937748/182892193-8accc301-33fb-4979-94af-28665fb2f579.png">
l

**Comment by LinusU** [comment]
I can't see that you are using `fs-temp` in any of the code posted, so your problem is likely in another package...

If you want further help you'll need to submit a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example).

**Comment by mercmobily** [comment]
Sorry to pester... but it really looks like fs-temp is not compatible with node 16.
This is the initial report: https://github.com/nodejs/node/issues/38058
I found it as I am having the exact same problem with Multer + fs-temp.

# Using node v16.20.2 (it DOESN'T work)

Painfully tracing the code, you get this:

````
// multer/lib/read-body.js:70:
target = temp.createWriteStream()
````

This takes you to `node:internal/fs/streams` line 311:

````
function WriteStream(path, options) {
  if (!(this instanceof WriteStream))
    return new WriteStream(path, options);

  options = copyObject(getOptions(options, kEmptyObject));

  // Only buffers are supported.
  options.decodeStrings = true;

  if (options.fd == null) {
    this.fd = null;
    this[kFs] = options.fs || fs;
    validateFunction(this[kFs].open, 'options.fs.open');

    // Path will be ignored when fd is specified, so it can be falsy
    this.path = toPathIfFileURL(path);
    this.flags = options.flags === undefined ? 'w' : options.flags;
    this.mode = options.mode === undefined ? 0o666 : options.mode;

    validatePath(this.path);
````

This part fails at the very last bit, `validatePath(this.path);`, since `this.path` is null. The problem is that _the code shouldn't be here in the first place_ since _`fd` should be set, but it's not.

 # Using node v12.22.12 (it WORKS)

As above, you get here:

````
// multer/lib/read-body.js:70:
target = temp.createWriteStream()
````
This takes you to `node:internal/fs/streams` line 264. here in Node 12 things look VASTLY different.
The function is completely different. But, the key issue is that here `fd` is set and things go smoothly.
 

# Thoughts

It seems to come down to the fact that node 16 doesn't like having a `null` path.
Looking at the [node API](https://nodejs.org/api/fs.html#fscreatewritestreampath-options):

> if fd is specified, [<fs.WriteStream>](https://nodejs.org/api/fs.html#class-fswritestream) will ignore the path argument and will use the specified file descriptor.

After more digging, I noticed that in node 21 the internal function  `WriteStream(path, options)` will do this:

````
 if (typeof this.fd !== 'number')
    this.open(); 
````

Whereas node 16 has NO such provision. So, `fd` stays null, and the check on the path fails (as it should).

I did some more digging, looking for why they removed `this.open()`.
It was removed [here](https://github.com/nodejs/node/commit/773769df60ac4f2448fa88b2ece035de2512928f)

Then the check for `fd` (and subsequent opening) was out altogether with [this commit](https://github.com/nodejs/node/commit/54b36e401d2b72d95e5f1dbbc787f6beed639347) . The commit refers to [this issue](https://github.com/nodejs/node/issues/23133) and it was a very major rework of how streams work.

The focus was to prevent this from being a problem:

````
const rec = db.getRecord(blockId)
pipeline(
  req, 
  fs.createWriteStream(blockPath, { flags: 'wx' })
    .on('open', () => {
      // Oops, this can be called after rec.destroy() :/.
      rec.set({
        path: blockPath
      })
    }),
  err => {
    rec.destroy()
    callback(err)
  }
)
````

One [key comment](https://github.com/nodejs/node/issues/23133#issuecomment-427726122) reads:

> I think we can make the fs stream not emit open at all if the stream is already destroyed.

I really hope this helps.

**Comment by LinusU** [comment]
This was fixed in [🚢 2.0.0 / 2021-07-24](https://github.com/LinusU/fs-temp/releases/tag/v2.0.0)

Specifically this commit: f02bfe3acfec76885b46fc91838b5edcaf1d356a

Are you able to update and try it again?
